### ISSUE TITLE
Intrepid2: Compile error

### ISSUE THREAD

**Issue by stanmoore1** [trilinos/Trilinos#6233]
## Bug Report
@trilinos/intrepid2

### Description
Compile error:
```
Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(112): error: a class-qualified name is required

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(112): error: expected a ";"

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(117): error: a class-qualified name is required

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(117): error: expected a ";"

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(122): error: a class-qualified name is required

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(122): error: expected a ";"

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(130): error: a class-qualified name is required

Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp(130): error: expected a ";"
...
```

### Steps to Reproduce
SHA1: 3f190fe, using gcc 7.3.1 on Broadwell hardware

**Comment by stanmoore1** [comment]
@mperego @kyungjoo-kim any ideas?

**Comment by kyungjoo-kim** [comment]
@CamelliaDPG Nate seems to change the typedef and introduce some macros.

**Comment by stanmoore1** [comment]
This works fine for CUDA but fails for gcc.

**Comment by CamelliaDPG** [comment]
Yes, this one will be my issue; this is my attempt to do deprecation warnings on typedefs.  There doesn't seem to be a very good way to do this in C++11 (IIRC, C++14 does provide something).  I'm guessing that gcc 7.3.1 is different from what's on the PR testing machines, since the PR tests for this succeeded.  Anyway, I can investigate.

**Comment by CamelliaDPG** [comment]
@stanmoore1, I have tried this on my local machine with gcc 7.4.0 (which is what I was able to install easily), and I'm not immediately reproducing.  It seems pretty weird to me that gcc 7.2.0 (on the PR testing machines) would work, and gcc 7.4.0 would work, but gcc 7.3.1 would not.  So I'm guessing that there's a sensitivity to the Trilinos configuration.  Would you please send me your Trilinos configure script?  I'll try to match your configuration as closely as I can.

**Comment by CamelliaDPG** [comment]
Fixed by #6244.
